### PR TITLE
Update the RetinaNet example

### PR DESCRIPTION
**Pull request by LukeWood**
Here are the final COCO metric results of this pipeline as evaluated by both KerasCV & MSCOCO:

Screen Shot 2022-10-18 at 12 06 23 PM

Here are sample detections:
download-1

closes: https://github.com/keras-team/keras-cv/issues/807

### CODE REVIEW

**Comment by bhack** [comment]
It is hard to understand the diff as you moved the file. What kind of changes we had?

**Comment by LukeWood** [comment]
> It is hard to understand the diff as you moved the file. What kind of changes we had?

migrated input pipeline away from the KPLs to @tanzhenyu 's from RCNN

**Comment by LukeWood** [comment]
@tanzhenyu

**Comment by tanzhenyu** [comment]
> It is hard to understand the diff as you moved the file. What kind of changes we had?

Maybe we could use `git mv files` so it's easier to check what happened exactly?

**Comment by LukeWood** [comment]
> > It is hard to understand the diff as you moved the file. What kind of changes we had?
> 
> Maybe we could use `git mv files` so it's easier to check what happened exactly?

hmm - I keep trying to do this but I can't get it to work.

**Comment by tanzhenyu** [comment]
> > > It is hard to understand the diff as you moved the file. What kind of changes we had?
> > 
> > 
> > Maybe we could use `git mv files` so it's easier to check what happened exactly?
> 
> hmm - I keep trying to do this but I can't get it to work.

You might need to git reset in order to do so. If this is too complicated, it's fine not to do it

**Comment by LukeWood** [comment]
@tanzhenyu I also updated the loss strings FYI

**Comment by LukeWood** [comment]
> > > > It is hard to understand the diff as you moved the file. What kind of changes we had?
> > > 
> > > 
> > > Maybe we could use `git mv files` so it's easier to check what happened exactly?
> > 
> > 
> > hmm - I keep trying to do this but I can't get it to work.
> 
> You might need to git reset in order to do so. If this is too complicated, it's fine not to do it

yeah, sorry about that - its basically an entirely new script at this point though.

File: examples/models/object_detection/retina_net/pascal_voc/train.py [hunk 1]
```diff
@@ -0,0 +1,567 @@
...
+)
+eval_ds = eval_ds.apply(
+    tf.data.experimental.dense_to_ragged_batch(8, drop_remainder=True)
+)
+
+
+def visualize_dataset(dataset, bounding_box_format, path=None):
+    example = next(iter(dataset))
+    images, boxes = example["images"], example["bounding_boxes"]
+    visualization.plot_bounding_box_gallery(
```
Comment (tanzhenyu) [inline examples/models/object_detection/retina_net/pascal_voc/train.py:253]: Is this part of keras_cv?

File: examples/models/object_detection/retina_net/pascal_voc/train.py [hunk 2]
```diff
@@ -0,0 +1,567 @@
...
+
+We'll use the KerasCV API to construct a RetinaNet model.  In this tutorial we use
+a pretrained ResNet50 backbone using weights.  In order to perform fine-tuning, we
+freeze the backbone before training.  When `include_rescaling=True` is set, inputs to
+the model are expected to be in the range `[0, 255]`.
+"""
+
+model = keras_cv.models.RetinaNet(
+    # number of classes to be used in box classification
+    classes=20,
```
Comment (tanzhenyu) [inline examples/models/object_detection/retina_net/pascal_voc/train.py:301]: len(class_ids)?

File: examples/models/object_detection/retina_net/pascal_voc/train.py [hunk 3]
```diff
@@ -0,0 +1,567 @@
...
+callbacks = [
+    keras.callbacks.TensorBoard(log_dir="logs"),
+    keras.callbacks.ReduceLROnPlateau(patience=5),
+    keras.callbacks.ModelCheckpoint(CHECKPOINT_PATH, save_weights_only=True),
+]
+
+history = model.fit(
+    train_ds,
+    validation_data=eval_ds.take(10),
+    epochs=100,
```
Comment (tanzhenyu) [inline examples/models/object_detection/retina_net/pascal_voc/train.py:353]: 100 epochs seem too much?

File: examples/models/object_detection/retina_net/pascal_voc/train.py [hunk 4]
```diff
@@ -0,0 +1,567 @@
...
+    font_scale=1,
+    class_mapping=class_mapping,
+    path="inference.png",
+)
+
+"""
+## Evaluation
+"""
+
+coco_suite = [
```
Comment (tanzhenyu) [inline examples/models/object_detection/retina_net/pascal_voc/train.py:418]: why not just model.fit(train_ds, validation_data=eval_ds)?

File: keras_cv/models/object_detection/retina_net/retina_net.py [hunk 5]
```diff
@@ -116,7 +116,6 @@ def __init__(
         classification_head=None,
         box_head=None,
         evaluate_train_time_metrics=False,
-        name="RetinaNet",
```
Comment (tanzhenyu) [inline keras_cv/models/object_detection/retina_net/retina_net.py:119]: why removing the name?

File: keras_cv/models/object_detection/retina_net/retina_net.py [hunk 6]
```diff
@@ -506,6 +503,36 @@ def _parse_backbone(backbone, include_rescaling, backbone_weights):
...
+def _parse_box_loss(loss):
+    if not isinstance(loss, str):
+        # support arbitrary callables
+        return loss
+
+    # case insensitive comparison
+    if loss.lower() == "smoothl1":
+        return keras_cv.losses.SmoothL1Loss(l1_cutoff=1.0, reduction="none")
+
+    raise ValueError(
```
Comment (tanzhenyu) [inline keras_cv/models/object_detection/retina_net/retina_net.py:515]: any particular reason that this cannot support Huber loss?

File: keras_cv/models/object_detection/retina_net/retina_net.py [hunk 7]
```diff
@@ -506,6 +503,36 @@ def _parse_backbone(backbone, include_rescaling, backbone_weights):
...
+
+
+def _parse_classification_loss(loss):
+    if not isinstance(loss, str):
+        # support arbitrary callables
+        return loss
+
+    # case insensitive comparison
+    if loss.lower() == "focal":
+        return keras_cv.losses.FocalLoss(from_logits=True, reduction="none")
```
Comment (tanzhenyu) [inline keras_cv/models/object_detection/retina_net/retina_net.py:528]: Do we check `from_logits` if the user pass a callable?

**Review by tanzhenyu — COMMENTED** [review]
Thanks for the PR!

File: examples/models/object_detection/retina_net/pascal_voc/train.py [hunk 8]
```diff
@@ -0,0 +1,567 @@
+import tensorflow as tf
+import tensorflow_datasets as tfds
+from luketils import visualization
```
Comment (innat) [inline examples/models/object_detection/retina_net/pascal_voc/train.py:3]: Some may not recognize `luketils`.

File: examples/models/object_detection/retina_net/pascal_voc/train.py [hunk 9]
```diff
@@ -0,0 +1,567 @@
...
+        return {"images": image, "bounding_boxes": bounding_boxes}
+
+    return apply
+
+
+train_ds = train_ds.map(
+    proc_train_fn("xywh", image_size), num_parallel_calls=tf.data.AUTOTUNE
+)
+train_ds = train_ds.apply(
+    tf.data.experimental.dense_to_ragged_batch(8, drop_remainder=True)
```
Comment (innat) [inline examples/models/object_detection/retina_net/pascal_voc/train.py:239]: `BATCH_SIZE`

File: examples/models/object_detection/retina_net/pascal_voc/train.py [hunk 10]
```diff
@@ -0,0 +1,567 @@
...
+    # and also causes a massive performance hit.  It can, however be useful to produce
+    # train time metrics when debugging your model training pipeline.
+    evaluate_train_time_metrics=False,
+)
+# Fine-tuning a RetinaNet is as simple as setting backbone.trainable = False
+model.backbone.trainable = False
+
+metrics = [
+    keras_cv.metrics.COCOMeanAveragePrecision(
+        class_ids=range(21),
```
Comment (innat) [inline examples/models/object_detection/retina_net/pascal_voc/train.py:325]: `range(len(class_ids)+1)`

File: examples/models/object_detection/retina_net/pascal_voc/train.py [hunk 11]
```diff
@@ -0,0 +1,567 @@
...
+    return apply
+
+
+eval_ds = tfds.load("voc/2007", split="test", with_info=False, shuffle_files=True)
+eval_ds = eval_ds.map(
+    proc_eval_fn("xywh", [640, 640, 3]), num_parallel_calls=tf.data.AUTOTUNE
+)
+eval_ds = eval_ds.apply(
+    tf.data.experimental.dense_to_ragged_batch(8, drop_remainder=True)
+)
```
Comment (innat) [inline examples/models/object_detection/retina_net/pascal_voc/train.py:563]: `BATCH_SIZE`

File: examples/models/object_detection/retina_net/pascal_voc/train.py [hunk 12]
```diff
@@ -0,0 +1,567 @@
...
+eval_ds = eval_ds.prefetch(2)
+
```
Comment (innat) [inline examples/models/object_detection/retina_net/pascal_voc/train.py:565]: Any reason to pick 2?
`...eval_ds.prefetch(tf.data.AUTOTUNE)`

File: examples/models/object_detection/retina_net/pascal_voc/train.py (hunk 8 above)
Comment (LukeWood) [reply examples/models/object_detection/retina_net/pascal_voc/train.py:3]: @tanzhenyu should we just add the visualization functions to keras_cv?  We can just copy paste them, but it requires OpenCV.   I originally had them here just to avoid copy-pasting them all over.

File: examples/models/object_detection/retina_net/pascal_voc/train.py (hunk 1 above)
Comment (LukeWood) [reply examples/models/object_detection/retina_net/pascal_voc/train.py:253]: Should we add it?  My thought was to upstream it at some point - but I wasn't quite ready for a full API review.

File: examples/models/object_detection/retina_net/pascal_voc/train.py (hunk 3 above)
Comment (LukeWood) [reply examples/models/object_detection/retina_net/pascal_voc/train.py:301]: Added early stopping callback

File: examples/models/object_detection/retina_net/pascal_voc/train.py (hunk 4 above)
Comment (LukeWood) [reply examples/models/object_detection/retina_net/pascal_voc/train.py:309]: if you evaluate the coco metrics during training after each epoch it takes way too long to run.

File: keras_cv/models/object_detection/retina_net/retina_net.py (hunk 5 above)
Comment (LukeWood) [reply keras_cv/models/object_detection/retina_net/retina_net.py:119]: Added back - wasn't sure if we wanted/needed default names.  No strog opinion

File: keras_cv/models/object_detection/retina_net/retina_net.py (hunk 6 above)
Comment (LukeWood) [reply keras_cv/models/object_detection/retina_net/retina_net.py:515]: Sure, why not!  added.

File: keras_cv/models/object_detection/retina_net/retina_net.py (hunk 7 above)
Comment (LukeWood) [reply keras_cv/models/object_detection/retina_net/retina_net.py:528]: No we don't, should we just accept loss objects?

File: examples/models/object_detection/retina_net/pascal_voc/train.py (hunk 1 above)
Comment (tanzhenyu) [reply examples/models/object_detection/retina_net/pascal_voc/train.py]: Yeah, I think it's definitely fine to achieve the minimal here: only train_ds without a COCO evaluation or visualizaiton. In the end we should have those, but given it's not fully flushed-out yet, better to exclude it for this PR.

File: keras_cv/models/object_detection/retina_net/retina_net.py (hunk 7 above)
Comment (tanzhenyu) [reply keras_cv/models/object_detection/retina_net/retina_net.py:534]: I think loss objects was added to Keras *after* loss callables, with intention to be the default, so yeah only accepting objects is fine with me

File: examples/models/object_detection/retina_net/pascal_voc/train.py (hunk 4 above)
Comment (tanzhenyu) [reply examples/models/object_detection/retina_net/pascal_voc/train.py:315]: Maybe we should have some canonical training workflow like:
```
model.compile(..., metrics=[cocomAP_suite])
model.fit(train_ds, validation_data=eval_ds, epochs=?)
model.evaluate(eval_ds)
```
Though with cocomAP accuracy still under investigation, it's definitely ok not to include them for now. No strong opinion here.

File: examples/models/object_detection/retina_net/pascal_voc/train.py (hunk 3 above)
Comment (tanzhenyu) [reply examples/models/object_detection/retina_net/pascal_voc/train.py:307]: Sounds good. No strong opinion whether you want to follow the paper strictly by epochs, or early stopping. (It's not easy to set the `patience` part for those callbacks though)

**Comment by LukeWood** [comment]
@tanzhenyu does this look good to you?

File: examples/models/object_detection/retina_net/pascal_voc/train.py [hunk 13]
```diff
@@ -0,0 +1,464 @@
...
+    classification_loss=keras_cv.losses.FocalLoss(from_logits=True, reduction="none"),
+    box_loss=keras_cv.losses.SmoothL1Loss(l1_cutoff=1.0, reduction="none"),
+    optimizer=optimizer,
+    metrics=coco_suite,
+)
+
+model.load_weights(CHECKPOINT_PATH)
+
+
+def proc_eval_fn(bounding_box_format, target_size):
```
Comment (tanzhenyu) [inline examples/models/object_detection/retina_net/pascal_voc/train.py:407]: Consider refactoring this with `proc_train_fn`? I didn't really use this other than in the eval.py, so I copied a lot of code for convenience

**Review by tanzhenyu — APPROVED** [review]

File: examples/models/object_detection/retina_net/pascal_voc/train.py (hunk 13 above)
Comment (LukeWood) [reply examples/models/object_detection/retina_net/pascal_voc/train.py:407]: I think most of this will go away with KPLs anyways - should we just focus our efforts there?
